File: crates/agent/src/step.rs
//! The pure step function (§11) — the hard wall between thinking and doing.
//! `step` cannot do I/O; it can only describe it, so everything here tests on
//! the host by asserting on the returned effects (ARCHITECTURE §5).

use kernel::{Event, EventKind};

use crate::ask::{call_model, config, scoped_tools};
use crate::components::{Observations, Task};
use crate::effect::Effect;
use crate::paper;
use crate::reply::{parse_reply, ParsedReply};
use crate::state::AgentState;
use crate::tools::ToolResult;
use crate::{answer, ending, stages, steer, stop, verify, window};

/// A state and the effects it wants run — what every transition below returns.
type Stepped = (AgentState, Vec<Effect>);

/// The frozen §11 signature. Owns ALL transitions: parse the event against the
/// current phase's contract, match the result against its exits, emit the next
/// phase's effects; malformed replies, illegal transitions and exhausted
/// budgets are handled by the machine, never by prose. Consumes and returns
/// state by value, which is what makes debugging a fold over the log.
pub fn step(state: AgentState, input: Event) -> Stepped {
    // NOTHING NEW AFTER A STOP (R16-P0-2): `advance` starts work only by
    // RETURNING an effect, so one check on the way out is the whole boundary.
    let (state, effects) = advance(state, input);
    stop::boundary(state, effects)
}

/// THE EXIT TABLE — one arm per fact the agent can be handed. Every arm is a
/// pure state transform returning effects; the arms with anything to explain
/// are functions below, so this reads as the list of things that can happen.
fn advance(mut state: AgentState, input: Event) -> Stepped {
    match input.kind {
        // Stop pressed: recorded, NOTHING emitted (`stop.rs` holds why). An
        // idle agent is already stopped, so this only takes on a running turn.
        EventKind::Custom { ref kind, .. } if kind == stop::STOP_REQUESTED => {
            state.stopping = state.task.is_some();
            (state, Vec::new())
        }
        EventKind::UserMessage { ref text, .. } if state.task.is_some() => {
            on_steer(state, text, input.at)
        }
        EventKind::UserMessage { text, .. } => on_task(state, text, input.at),
        EventKind::ModelReplied { text, .. } if state.compacting => {
            on_summary(state, &text, input.at)
        }
        EventKind::Custom { ref kind, .. }
            if kind == "core.compaction_failed" && state.compacting =>
        {
            on_compaction_failed(state, input.at)
        }
        // The completed reply (ADR-002: deltas never reach the log): either
        // tool calls to run, or the answer that ends the turn.
        EventKind::ModelReplied { text, .. } => on_reply(state, &text, input.at),
        // One tool came back. The batch is done when the last result lands,
        // and then — and only then — the model sees them all.
        EventKind::ToolInvoked { tool, ok, output, .. } => {
            let result = ToolResult {
                tool: tool.0,
                ok,
                output: output.clone(),
                error: output,
            };
            on_tool_result(state, &result, input.at)
        }
        // Facts observed but not acted on: quiescence, not effects.
        _ => (state, Vec::new()),
    }
}

/// A user utterance DURING a turn is steering, not a new turn. The naive
/// reading — reset the counters and call the model — would ask the model twice
/// at once and then decrement `pending_tools` below the batch still in flight.
/// So the sentence is appended and NO WORK is emitted: the round already
/// running finishes, and the next `call_model` assembles a paper with the
/// interjection in it. …AND IT SAYS SO IN THE LOG (R18-P0-1); `steer.rs` why.
fn on_steer(mut state: AgentState, text: &str, at: kernel::Timestamp) -> Stepped {
    paper::push_history(&mut state.paper, "user", text, at);
    state.steered = true;
    (state, vec![steer::carried()])
}

/// A user utterance starts (or redirects) the turn: assemble the paper under
/// the phase's budget and ask. The Document rides the effect (I13).
///
/// Everything a turn counts is reset, `stopping` included — a stop ends ONE
/// turn, not the next — and the cursor with them (20): a turn starts at the
/// first stage its file declares. Compaction runs at the TOP of a turn, with
/// the question just asked already in the window (Python `_step`): summarise
/// the question away and the model answers one it cannot see.
fn on_task(mut state: AgentState, text: String, at: kernel::Timestamp) -> Stepped {
    state.task = Some(text.clone());
    paper::set_component(&mut state.paper, &Task { text: text.clone() }, at);
    paper::push_history(&mut state.paper, "user", &text, at);
    (state.pending_tools, state.tool_rounds, state.stopping) = (0, 0, false);
    verify::clear(&mut state);
    let opened = stages::open(&mut state, at);
    let effect = window::compaction(&mut state, at).unwrap_or_else(|| call_model(&mut state, at));
    (state, opened.into_iter().chain([effect]).collect())
}

/// The summarisation FAILED. A compaction costs a compaction and never a
/// conversation (Python `compact`: warned about, carried on from), so the
/// window is left as it was and the turn the person asked for is taken now, in
/// full — it was never asked at all before (09 walk).
fn on_compaction_failed(mut state: AgentState, at: kernel::Timestamp) -> Stepped {
    state.compacting = false;
    let effect = call_model(&mut state, at);
    (state, vec![effect])
}

/// The summary is back. It REPLACES the older window, and the turn the person
/// asked for is taken against the compacted paper. The reply is the
/// summarizer's, which is why it never becomes an answer.
fn on_summary(mut state: AgentState, text: &str, at: kernel::Timestamp) -> Stepped {
    state.compacting = false;
    if window::compacted(&mut state.paper, text, state.keep_recent, at) {
        state.compactions += 1;
    }
    let effect = call_model(&mut state, at);
    (state, vec![effect])
}

/// One reply against the phase's contract. Tool calls act; anything else goes
/// to `answer.rs`, which owns the turn's cheap exit and the gate in front of it.
fn on_reply(mut state: AgentState, text: &str, at: kernel::Timestamp) -> Stepped {
    let cfg = config(&state);
    let batches = match parse_reply(cfg.contract, text) {
        Ok(ParsedReply::Tools(batches)) if !batches.is_empty() => batches,
        _ => return answer::answered(state, text, at),
    };
    paper::push_history(&mut state.paper, "assistant", text.trim(), at);
    // Batch ORDER is the schedule: a later line's calls are emitted after an
    // earlier line's, and the model sees no result until every call it wrote
    // has come back. The batch INDEX rides out on each delegation, so the
    // runtime can run one line's sub-agents concurrently in their own Workers
    // — the other half of the layout rule (increment 06).
    let tools = scoped_tools(&state, &cfg);
    let effects: Vec<Effect> = batches
        .into_iter()
        .enumerate()
        .flat_map(|(line, calls)| {
            let tools = &tools;
            calls
                .into_iter()
                .map(move |call| crate::subagent::invoke_or_refuse(tools, call, line as u16))
        })
        .collect();
    state.pending_tools = effects.len();
    (state, effects)
}

/// One tool result. The batch is not done until the last one lands; then the
/// model gets them all at once and the loop goes round — up to `max_rounds`.
fn on_tool_result(mut state: AgentState, r: &ToolResult, at: kernel::Timestamp) -> Stepped {
    paper::push_history(&mut state.paper, "Result", &r.line(), at);
    let seen = Observations { lines: vec![r.line()] };
    paper::set_component(&mut state.paper, &seen, at);
    // THE FOLD, in log order — a write clears the flag, a command that printed
    // something after it sets it. `verify.rs` holds why order is enough.
    verify::observe(&mut state, &r.tool, r.ok, &r.output);
    state.pending_tools = state.pending_tools.saturating_sub(1);
    if state.pending_tools > 0 {
        return (state, Vec::new());
    }
    state.tool_rounds += 1; // a round is COMPLETE when its last result lands
    round_again(state, at)
}

/// The round is complete: go round once more, or stop at the ceiling. A steer
/// is consumed either way — asked again with the sentence in front of it, or
/// the turn ends and the sentence opens the next one.
///
/// THE CEILING IS AN ENDING LIKE ANY OTHER (R17-P0-2). It used to emit a
/// `core.note`, the kind the machine uses for anything it says, so only the
/// conversation could tell this ending from an answer. Its wording moved to
/// `core::failure::ending` beside the others', off that one fact.
///
/// Compaction runs before EVERY round, not only at the top of a turn: a turn
/// may take sixty-four (15C), each round appending a reply and a result, and
/// `assemble` degrades silently at the budget — so the late rounds of a long
/// run were quietly losing the task they were working on.
fn round_again(mut state: AgentState, at: kernel::Timestamp) -> Stepped {
    if state.tool_rounds >= state.max_rounds {
        let effect = ending::end(&mut state, ending::ROUND_CEILING);
        return (state, vec![effect]);
    }
    state.steered = false;
    let effect = window::compaction(&mut state, at).unwrap_or_else(|| call_model(&mut state, at));
    (state, vec![effect])
}

File: packages/agent/src/step.js
/**
 * THE PURE STEP FUNCTION — the wall between thinking and doing, and the ONLY
 * writer of an `AgentState`.
 *
 * It takes the state and one arriving fact and returns a NEW state plus the
 * effects it wants run; it cannot do I/O, so every claim about the loop is a
 * claim about a returned array and tests on the host (I3, I7). The predecessor
 * had this shape too and then broke it from outside — two sites cleared
 * `agent.task` without going through a transition, which is exactly how a
 * result could arrive against a turn the reducer still believed was running.
 * There is one door. An out-of-band mutation of the snapshot handed in throws.
 *
 * THE ORDER OF THE THREE THINGS THIS DOES. Refuse what does not belong to the
 * live turn (I21) — then act — then, on the way out, honour a pressed Stop at
 * the single boundary every arm returns through.
 * @module
 */

import { MODEL_ENDPOINT } from '@harness/kernel'
import { callModel, invokeTool } from './effect.js'
import { ANSWERED, ROUND_CEILING, RESPOND, endTurn, endingFor } from './ending.js'
import { boundary, isStopRequest } from './stop.js'
import { carried } from './steer.js'
import { dropped, refusal } from './turn.js'

/** @typedef {import('@harness/kernel').Timestamp} Timestamp */
/** @typedef {import('./effect.js').Effect} Effect */
/** @typedef {import('./state.js').AgentState} AgentState */
/** @typedef {import('./turn.js').Incoming} Incoming */
/** @typedef {import('./turn.js').Reply} Reply */

/** A state and the effects it wants run — what every transition returns. @typedef {{state: AgentState, effects: Effect[]}} Stepped */

/**
 * The frozen signature. Owns ALL transitions.
 * @param {AgentState} state a snapshot; it is never written to
 * @param {Incoming} incoming one fact, and the turn it was produced under
 * @returns {Stepped}
 */
export function step(state, incoming) {
  const why = refusal(state, incoming)
  if (why) return { state, effects: [dropped(state, incoming, why)] }
  const taken = advance(state, incoming)
  return boundary(taken.state, taken.effects)
}

/**
 * THE EXIT TABLE — one line per fact the agent can be handed, so this reads as
 * the list of things that can happen to a turn. Everything with a reason to
 * give is a function below.
 * @param {AgentState} state @param {Incoming} incoming @returns {Stepped}
 */
function advance(state, incoming) {
  const fact = incoming.fact
  // Stop pressed: recorded, NOTHING emitted. An idle agent is already stopped.
  if (isStopRequest(fact)) return { state: { ...state, stopping: state.task !== null }, effects: [] }
  if (fact.type === 'user_message' && state.turnId !== '') return onSteer(state)
  if (fact.type === 'user_message') return onTask(state, fact.text, incoming)
  if (fact.type === 'model_replied') return onReply(state, incoming)
  if (fact.type === 'tool_invoked') return onToolResult(state, fact)
  // Facts observed but not acted on: quiescence, not effects.
  return { state, effects: [] }
}

/**
 * A user utterance DURING a turn is steering. The sentence is appended to the
 * log by the driver and NO work is emitted: the round already running finishes,
 * and the next model call assembles a paper with the interjection in it. The
 * naive reading — reset the counters and call the model — asks the model twice
 * at once and then counts the batch in flight down through a fresh counter.
 * @param {AgentState} state @returns {Stepped}
 */
function onSteer(state) {
  return { state: { ...state, steered: true }, effects: [carried()] }
}

/**
 * A user utterance starts the turn. Everything a turn counts is reset,
 * `stopping` included — a stop ends one turn, not the next.
 *
 * A MESSAGE WITH NO TURN ID IS AN ANOMALY, not an occasion to mint one: ids are
 * injected (I7), the spine mints one per accepted message, and a turn that
 * named itself would be a turn no effect could be matched against.
 * @param {AgentState} state @param {string} text @param {Incoming} incoming @returns {Stepped}
 */
function onTask(state, text, incoming) {
  if (!incoming.turnId) {
    return { state, effects: [dropped(state, incoming, 'it arrived with no turn to run it under')] }
  }
  /** @type {AgentState} */
  const turn = {
    ...state,
    task: text, turnId: incoming.turnId, awaiting: 'model',
    pendingTools: 0, toolRounds: 0, observations: [], steered: false, stopping: false,
  }
  return { state: turn, effects: [nextCall(turn)] }
}

/**
 * ONE REPLY, READ AS A SIGNAL. Calls are work; no calls is an ending, and WHICH
 * ending is the provider's `finish` — never the shape of the prose.
 *
 * A reply with no signal at all is refused rather than guessed at: "no call in
 * this text" meaning "the model answered" is the defect this whole rewrite of
 * the arm exists to remove.
 *
 * CALLS BEAT THE SIGNAL where a provider sends both `stop` and a call — the
 * model asked for work, and the signal only decides how a call-less reply
 * ended.
 * @param {AgentState} state @param {Incoming} incoming @returns {Stepped}
 */
function onReply(state, incoming) {
  const reply = incoming.reply
  if (!reply) {
    return { state, effects: [dropped(state, incoming, 'the reply carried no ending signal, and a turn ends on a signal and never on silence')] }
  }
  if (reply.calls.length === 0) return endTurn(state, endingFor(reply.finish))
  if (reply.calls.some((call) => call.tool === RESPOND)) return endTurn(state, ANSWERED)
  /** @type {AgentState} */
  const acting = { ...state, awaiting: 'tools', pendingTools: reply.calls.length, observations: [] }
  return { state: acting, effects: reply.calls.map((call) => invokeTool(state.turnId, call.tool, call.args)) }
}

/**
 * ONE TOOL RESULT, APPENDED. The batch is not done until the last one lands,
 * and only then does the model see them — that is what makes one LINE of calls
 * one observation.
 *
 * `observations` IS AN ARRAY. The Rust built a one-line `Observations`
 * component per result and upserted it by id, so three calls on one line
 * produced three overwrites and the model saw the last one. Each result is
 * appended in arrival order, and the round that follows starts a new list.
 * @param {AgentState} state @param {{tool: string, ok: boolean, output: string}} result @returns {Stepped}
 */
function onToolResult(state, result) {
  /** @type {AgentState} */
  const seen = {
    ...state,
    observations: [...state.observations, resultLine(result)],
    pendingTools: state.pendingTools - 1,
  }
  if (seen.pendingTools > 0) return { state: seen, effects: [] }
  const round = { ...seen, toolRounds: seen.toolRounds + 1 } // a round is COMPLETE when its last result lands
  if (round.toolRounds >= round.maxRounds) return endTurn(round, ROUND_CEILING)
  /** @type {AgentState} */
  const asking = { ...round, awaiting: 'model', steered: false }
  return { state: asking, effects: [nextCall(asking)] }
}

/**
 * One result as the model reads it. FAILURE IS IN THE LINE: the Rust rendered
 * `tool: output` for both outcomes and carried `ok` no further, so a model
 * could tell a failed call from a successful one only by reading the prose the
 * failure happened to contain (I16).
 * @param {{tool: string, ok: boolean, output: string}} result @returns {string}
 */
function resultLine(result) {
  return result.ok ? `${result.tool}: ${result.output}` : `${result.tool} failed: ${result.output}`
}

/**
 * ASK THE MODEL. The Document rides the effect, because nothing reaches a model
 * except as an assembled Document (I13).
 *
 * THE ASSEMBLY IS NOT HERE YET and this says so rather than pretending: the
 * paper's sources cross as the document's sections verbatim. `assemble` (lane
 * A) puts the budget, the degrade ladder and the affordances between them, and
 * `ask.js` (B11) is the seam that will hold both — this is the one call site it
 * replaces, not a second door beside it.
 * @param {AgentState} state @returns {Effect}
 */
function nextCall(state) {
  return callModel({
    turnId: state.turnId,
    document: { sections: state.paper.sources },
    format: { target: 'openai', vision: false, audio: false },
    endpoint: MODEL_ENDPOINT,
    model: state.model,
    temperature: state.temperature,
  })
}
